src/libAtomVM/context.c

```diff
@@ -404,10 +404,12 @@ static struct ResourceMonitor *context_monitors_handle_terminate(Context *ctx)
                         AVM_ABORT();
                     }
 
+                    term exited_pid = term_from_local_process_id(ctx->process_id);
+                    context_unlink(target, exited_pid);
```

Comment:
OK. I will check the test cases.
I guess it is in the calling function `context_destroy` that the table is locked?
I also noticed that in nifs.c in function `nif_erlang_exit`, there is a similar sequence. Maybe that needs update too?

Reply:
The lock is acquired in `context_destroy` by calling `synclist_wrlock` function. For now, there is a single lock for every processes in the table.
https://github.com/atomvm/AtomVM/blob/cddb5fc0ba6e01a26c5ed9d744f6367ba3b8619c/src/libAtomVM/context.c#L127

The lock is also released after the static function `context_monitors_handle_terminate` is called.
https://github.com/atomvm/AtomVM/blob/cddb5fc0ba6e01a26c5ed9d744f6367ba3b8619c/src/libAtomVM/context.c#L139

We had a bug in this sequence at some point.

`nif_erlang_exit` looks OK to me as it uses the "regular" sequence with `globalcontext_get_process_lock`, `mailbox_send_*` and `globalcontext_get_process_unlock`, which would automatically work if we changed the locking implementation.

`nif_erlang_link` does call `context_link` (and maybe `context_unlink`) on the target, and you will notice it also does so after calling `globalcontext_get_process_lock` on target_pid. There is no need to lock with `ctx` because there is only one lock (and it would deadlock).

Reply:
Sorry, I think I was a bit unclear to what to update.
The locks looks ok, but in nif_erlang_exit you do not call context_unlink to remove the connection to the exited process in the target process. This is what I did in `context_monitors_handle_terminate` and maybe `nif_erlang_exit` should be updated the same way with a call to `context_unlink(target, exited_pid) `?

Reply:
Sorry for my misunderstanding.

Your fix looks correct in `context_monitors_handle_terminate` because, when run for terminating process A, as B is in the list of processes linked to A, you know that A is also in the list of processes linked to B. So you can call context_unlink(B, A).

In `nif_erlang_exit` on the contrary to what the name implies, no process is exiting. If it's `exit/1`, an exception will be raised and the process may terminate at some point and `context_monitors_handle_terminate` will be called. If it's `exit/2` a kill signal may be sent, but then the process will be terminated later and `context_monitors_handle_terminate` will be called as well. There is no need to remove any link here, it will be done later.

FYI, monitors are not cleaned up as mentioned in the TODO comment.